Feature engineering slides (#127)
tidymodels/workshops: slides/advanced-01-feature-engineering.qmd

:::

## Your turn {transition="slide-in"}

![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}

*Create a `recipe()` for the hotel data to :*

-   *use a Yeo-Johnson (YJ) transformation on `lead_time`*
-   *convert factors to indicator variables*
-   *remove zero-variance variables*

```{r}
#| label: your-turn-make-recipe
#| echo: false
countdown(minutes = 3, id = "make-recipe")
```

## Minimal recipe `r hexes("recipes")` 

```{r}
#| label: hotel-rec
hotel_indicators <-
  recipe(avg_price_per_room ~ ., data = hotel_tr) %>% 
  step_YeoJohnson(lead_time) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_zv(all_predictors())
```

## Using a workflow `r hexes("recipes", "workflows", "parsnip", "tune")` {.annotation}

```{r}
#| label: lm-model
#| cache: true

set.seed(9)

hotel_lm_wflow <-
  workflow() %>%
  add_recipe(hotel_indicators) %>%
  add_model(linear_reg())
 
ctrl <- control_resamples(save_pred = TRUE)
hotel_lm_res <-
  hotel_lm_wflow %>%
  fit_resamples(hotel_rs, control = ctrl)

collect_metrics(hotel_lm_res)
```

## Your turn {transition="slide-in"}

![](images/parsnip-flagger.jpg){.absolute top="0" right="0" width="150" height="150"}

> *Use `fit_resamples()` to fit your workflow with a recipe.*

Reviewer: The learners will not necessarily have made their own workflow at this point. Are they supposed to use the one from the previous slides or pick a (different) model to make their own workflow?

Author: Maybe I don't understand; we did [make one on the previous slide](https://github.com/tidymodels/workshops/pull/127/files#diff-757f60aaa2bcbca2701235fb4c5e1184fc52674ac8a7d67ea7e79fd5838fcd4bR462)

*Collect the predictions from the results.*

```{r}
#| label: your-turn-resample-recipe
#| echo: false
countdown(minutes = 5, id = "resample-recipe")
```

## Holdout predictions `r hexes("recipes", "workflows", "parsnip", "tune")`

```{r}
#| label: goldout-pred
# Since we used `save_pred = TRUE`
lm_val_pred <- collect_predictions(hotel_lm_res)
lm_val_pred %>% slice(1:7)
```

## Calibration Plot `r hexes("probably")`

```{r}
#| label: lm-cal-plot
#| out-width: 40%
#| fig-width: 5
#| fig-height: 5
#| fig-align: "center"

library(probably)

cal_plot_regression(hotel_lm_res, alpha = 1 / 5)
```

> ## What do we do with the agent and company data? 

Author: I'll make "extra" slides using some of these slides plus the effect encoding steps and workflow sets

-   use effect encoding to replace the `agent` and `company` columns with the estimated effect of that predictor (in the extra materials)

```{r}
#| label: effects-calcs 
#| include: false

agent_stats <- 
  hotel_tr %>%
  group_by(agent) %>%
  summarize(
    ADR = mean(avg_price_per_room), 
    num_reservations = n(),
    .groups = "drop"
    ) %>%
  mutate(agent = reorder(agent, ADR))
```

## Per-agent statistics 

::: columns
::: {.column width="50%"}
```{r}
#| label: effects-freq 
#| echo: false
#| out-width: '90%'
#| fig-width: 6
#| fig-height: 3
#| fig-align: 'center'
#| dev: 'svg'
#| dev-args: list(bg = "transparent")
  
agent_stats %>%   
  ggplot(aes(x = num_reservations)) +
  geom_histogram(bins = 30, col = "blue", fill = "blue", alpha = 1/3) +
  labs(x = "Number of reservations per agent")
```
:::

::: {.column width="50%"}
```{r}
#| label: effects-adr 
#| echo: false
#| out-width: '90%'
#| fig-width: 6
#| fig-height: 3
#| fig-align: 'center'
#| dev: 'svg'
#| dev-args: list(bg = "transparent")

agent_stats %>%   
  ggplot(aes(x = ADR)) +
  geom_histogram(bins = 30, col = "red", fill = "red", alpha = 1/3) +
  labs(x = "Average ADR per agent")
```
:::
:::

> ## Collapsing factor levels

Reviewer: From here on out we don't have hexes in the corner anymore

> There is a recipe step that will redefine factor levels based on the their frequency in the training set: 

Reviewer: ```suggestion
There is a recipe step that will redefine factor levels based on their frequency in the training set: 
```

```{r}
#| label: step-other

> #| code-line-numbers: "3:4"

Reviewer: ```suggestion
#| code-line-numbers: "4"
```

hotel_other_rec <-
  recipe(avg_price_per_room ~ ., data = hotel_tr) %>% 
  step_YeoJohnson(lead_time) %>%
  step_other(agent, threshold = 0.001) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_zv(all_predictors())
```

```{r}
#| label: other-res 
#| include: false

retained_agents <-
  recipe(avg_price_per_room ~ ., data = hotel_tr) %>%
  step_mutate(original = agent) %>% 
  step_other(agent, threshold = 0.001) %>%
  step_dummy(all_nominal_predictors()) %>%
  step_zv(all_predictors()) %>% 
  prep() %>% 
  tidy(number = 2)

num_agents <- length(unique(hotel_tr$agent))
num_other <- num_agents - length(retained_agents$retained)
```

Using this code, `r num_other` agents (out of `r num_agents`) were collapsed into "other" based on the training set.

We _could_ try to optimize the threshold for collapsing (see the next set of slides on model tuning).

## Does othering help? 

```{r}
#| label: update-recipe
#| cache: true
#| code-line-numbers: "3|"
hotel_other_wflow <-
  hotel_lm_wflow %>%
  update_recipe(hotel_other_rec)

hotel_other_res <-
  hotel_other_wflow %>%
  fit_resamples(hotel_rs, control = ctrl)

collect_metrics(hotel_other_res)
```

> A little better MAE and much faster to complete.  

Reviewer: The code output does not actually show the MAE (and the RMSE is worse).

Now let's look at a more sophisticated tool called effect feature hashing. 

## Feature Hashing

Between `agent` and `company`, simple dummy variables would create `r length(unique(hotel_tr$agent)) + length(unique(hotel_tr$company))` new columns (that are mostly zeros).

Another option is to have a binary indicator that combines some levels of these variables.

> Feature hashing ([_FES_](https://bookdown.org/max/FES/encoding-predictors-with-many-categories.html),[_SMLTAR_](https://smltar.com/mlregression.html#case-study-feature-hashing), [_TMwR_](https://www.tmwr.org/categorical.html#feature-hashing)):  

Reviewer: ```suggestion
Feature hashing (for more see [_FES_](https://bookdown.org/max/FES/encoding-predictors-with-many-categories.html), [_SMLTAR_](https://smltar.com/mlregression.html#case-study-feature-hashing), and [_TMwR_](https://www.tmwr.org/categorical.html#feature-hashing)):  
```

> - uses the character value of the levels 
> - converts them to integer hash value
> - uses the integer to assign them to a specific indicator column. 

Reviewer: this is a bit of mix between singular and plural, would this be correct:
```suggestion
- uses the character values of the levels 
- converts them to integer hash values
- uses the integers to assign them to a specific indicator column. 
```

## Feature Hashing

Suppose we want to use 32 indicator variables for `agent`. 

For a agent with value "`Max_Kuhn`", a hashing function converts it to an integer (say `r strtoi(substr(rlang::hash("Max_Kuhn"), 26, 32), 16)`). 

To assign it to one of the 32 columns, we would use modular arithmetic to assign it to a column: 

```{r}
#| label: hash

> # For "Max_Kuhn" but a '1' in column: 

Reviewer: ```suggestion
# For "Max_Kuhn" put a '1' in column: 
```

210397726 %% 32
```

> [Hash functions](https://www.metamorphosite.com/one-way-hash-encryption-sha1-data-software) are used in cryptology and are meant to _emulate_ randomness. 

Reviewer 2: while this is true, the hashing function used in feature hashing are not cryptologically safe, rather they are fast and uniform.

Author: Good point.

> ## Feature Hashing Pros and Cons

Reviewer: Maybe group the pros and cons?

- The procedure will automatically work on new values of the predictors.

> - The is no real logic behind which factor levels are combined. 

Reviewer: ```suggestion
- There is no real logic behind which factor levels are combined. 
```

- It is fast. 
- We don't know how many columns to add (more in the next section).
- Some columns may have all zeros. 

> - "Signed" hashes add a sign to help avoid aliasing. 

Reviewer: This doesn't really feel like a pro or con

Author: It's a pro since the signs make it slightly more possible to differentiate between confounded levels. I'll add a note.

- If a indicator column is important to the model, we can't easily determine why. 

> ## Feature Hashing in recipes

Reviewer: Let's add the hex for textrecipes!

The textrecipes package has a step that can be added to the recipe: 

```{r}
#| label: hash-rec
#| code-line-numbers: "6-8|"
library(textrecipes)

hash_rec <-
  recipe(avg_price_per_room ~ ., data = hotel_tr) %>%
  step_YeoJohnson(lead_time) %>%

>   # Defaults to 32 signed indicator columns

Author: I'll merge the existing PRs ass-is (once review is done) and then go back and add tuning for these two pre-processing steps.

>   step_dummy_hash(agent) %>%
>   step_dummy_hash(company) %>%

Reviewer: ```suggestion
  textrecipes::step_dummy_hash(agent) %>%
  textrecipes::step_dummy_hash(company) %>%
```
like we did earlier with `embed::step_umap()`

  # Regular indicators for the others
  step_dummy(all_nominal_predictors()) %>% 
  step_zv(all_predictors())

hotel_hash_wflow <-
  hotel_lm_wflow %>%
  update_recipe(hash_rec)
```

## Feature Hashing in recipes

```{r}
#| label: hash-res
#| cache: true

hotel_hash_res <-
  hotel_hash_wflow %>%
  fit_resamples(hotel_rs, control = ctrl)

collect_metrics(hotel_hash_res)
```

About the same performance but now we can handle new values. 

## Debugging a recipe

- Typically, you will want to use a workflow to estimate and apply a recipe.

. . .

- If you have an error and need to debug your recipe, the original recipe object (e.g. `hash_rec`) can be estimated manually with a function called `prep()`. It is analogous to `fit()`. See [TMwR section 16.4](https://www.tmwr.org/dimensionality.html#recipe-functions)

. . .

- Another function (`bake()`) is analogous to `predict()`, and gives you the processed data back.

. . .

- The `tidy()` function can be used to get specific results from the recipe.

## Example

```{r}
#| label: prep-tidy-bake

> #| eval: false

Reviewer: Is this intentional?

Author: Yes. It is like that in the previous versions (I think because there is a lot of potential output).
